shared-libs/lineage/src/hydration.js

```diff
@@ -0,0 +1,387 @@
+      }
+      while (parent) {
+        if (parent._id) {
+          ids.push(parent._id);
+        }
+        parent = parent.parent;
+      }
+    });
+    return _.uniq(ids);
+  };
+
+  // for data_records, doesn't include the first-level contact (it counts as a parent).
+  const collectLeafContactIds = function(partiallyHydratedDocs) {
+    const ids = [];
+    partiallyHydratedDocs.forEach(function(doc) {
+      let current = doc;
+      if (current.type === 'data_record') {
+        current = current.contact;
+      }
+      while (current) {
+        const contactId = current.contact && current.contact._id;
+        if (contactId) {
+          ids.push(contactId);
+        }
+        current = current.parent;
+      }
+    });
+    return _.uniq(ids);
+  };
+
+  const fetchDocs = function(ids) {
+    if (!ids || !ids.length) {
+      return Promise.resolve([]);
+    }
+    const keys = _.uniq(ids.filter(id => id));
+    if (keys.length === 0) {
+      return Promise.resolve([]);
+    }
+
+    return DB.allDocs({ keys, include_docs: true })
+      .then(function(results) {
+        return results.rows
+          .map(function(row) {
+            return row.doc;
+          })
+          .filter(function(doc) {
+            return !!doc;
+          });
+      });
+  };
+
+  const hydrateDocs = function(docs) {
+    if (!docs.length) {
+      return Promise.resolve([]);
+    }
+    
+    const fetchPatientUuids = docs => Promise.all(docs.map(doc => fetchPatientUuid(doc)));
+    
+    const hydratedDocs = JSON.parse(JSON.stringify(docs)); // a copy of the original docs which we will incrementally hydrate and return
+    const knownDocs = [...hydratedDocs]; // an array of all documents which we have fetched
+
+    let patientUuids, patientDocs;
+    return fetchPatientUuids(hydratedDocs)
+      .then(function(uuids) {
+        patientUuids = uuids;
+        return fetchDocs(patientUuids);
+      })
+      .then(function(patients) {
+        patientDocs = patients;
+        knownDocs.push(...patients);
+        
+        const firstRoundIdsToFetch = _.uniq([
+          ...collectParentIds(hydratedDocs),
+          ...collectLeafContactIds(hydratedDocs),
+          
+          ...collectParentIds(patientDocs),
+          ...collectLeafContactIds(patientDocs),
+        ]);
+
+        return fetchDocs(firstRoundIdsToFetch);
+      })
+      .then(function(firstRoundFetched) {
+        knownDocs.push(...firstRoundFetched);
+        const secondRoundIdsToFetch = collectLeafContactIds(firstRoundFetched).filter(id => !knownDocs.some(doc => doc._id === id));
+        return fetchDocs(secondRoundIdsToFetch);
+      })
+      .then(function(secondRoundFetched) {
+        knownDocs.push(...secondRoundFetched);
+        
+        hydratedDocs.forEach((doc, i) => {
+          const reconstructLineage = (ids, docs) => {
+            return ids.map(id => {
+              // how can we use hashmaps?
+              return docs.find(doc => doc._id === id);
```

Comment:
When you're hydrating a contact that is the primary contact of it's parent or a report submitted about such a person, you'll end up with a circular object, which is something that we've been trying to avoid. For example: https://github.com/medic/medic/issues/4319

Reply:
So this would be solved here if I updated `extractParentIds` to guard against an infinite loop?

Reply:
I think the ideal result would be that you could `JSON.parse(JSON.stringify(hydratedDoc))` and get an equivalent object.

Reply:
I'm a bit confused. I'll ping you on Monday about this. Seems you're not specifically worried about an infinite loop in this code, but on this code causing infinite loops elsewhere. Confirming you have seen [this](https://github.com/medic/medic/pull/5926/files#diff-99a241955cc1d036a0d602cbafcfdb92R307) already?

Reply:
Should be much better now. Check it out.